### ISSUE TITLE
JsonObject doesn't report duplicate keys until requested

### ISSUE THREAD

**Issue by gregsdennis** [dotnet/runtime#70604]
### Description

A `JsonObject` will throw an `ArgumentException` when attempting to add a key which already exists.

```c#
var json = new JsonObject
{
	{ "foo", 1 },
	{ "bar", 2 },
	{ "foo", 3 },  // throws
};
```

However, parsing text with a duplicate key is accepted.

```c#
var json = JsonNode.Parse("{\"foo\":1, \"bar\":2, \"foo\":3}") as JsonObject;
```

It only throws when attempting to access the data

```c#
var foo = json["foo"]; // throws
var bar = json["bar"]; // also throws
```

### Reproduction Steps

See above

### Expected behavior

I expect failure earlier, on the parse.  This would indicate that it is the source data that is the problem.  Not throwing during the parse indicates that the parse completed successfully.

### Actual behavior

See above

### Regression?

Unsure.

### Known Workarounds

No known workarounds.

### Configuration

I've tested this in a .Net 5 test project so far, but I have no reason to believe the behavior is any different in other versions.

**Comment by skyoxZ** [comment]
https://www.ecma-international.org/publications-and-standards/standards/ecma-404/ (from [json.org](https://www.json.org/json-en.html))
> The JSON syntax does not impose any restrictions on 
the strings used as names, does not require that name strings be unique, and does not assign any 
significance to the ordering of name/value pairs. These are all semantic considerations that may be defined by 
JSON processors or in specifications defining specific uses of JSON for data interchange.

**Comment by eiriktsarpalis** [comment]
This appears to be by-design behavior, `JsonObject` will lazily initialize its underlying dictionary (and fail if duplicate keys are detected). I suspect changing this now might be a breaking change, but we might consider exposing an option in `JsonNodeOptions` forcing eager initialization of instances wrapping around `JsonElement`.

**Comment by gregsdennis** [comment]
> forcing eager initialization of instances wrapping around `JsonElement`.

Is that what results from a parse?  It's just an element under the hood?

**Comment by eiriktsarpalis** [comment]
> > forcing eager initialization of instances wrapping around `JsonElement`.
> 
> Is that what results from a parse? It's just an element under the hood?

Yes, any `JsonNode` parsing directly from JSON will simply wrap `JsonElement`:
https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonNode.Parse.cs#L68-L80

**Comment by skyoxZ** [comment]
Yes a node with duplicated key is invalid, but what if the node is not used? 
Please consider the code below. I think lazy Initialization may help improve performance when some nodes are never used and a bad node will not ruin whole json.
```csharp
string jsonString = "{ 'badNode': {'foo': 1, 'bar': 2, 'foo': 3 }, 'goodNode': { 'foo': 4 } }".Replace('\'', '"');
JsonObject json = JsonNode.Parse(jsonString).AsObject();
Console.WriteLine(json["goodNode"]["foo"]);
```

**Comment by gregsdennis** [comment]
> Yes a node with duplicated key is invalid, but what if the node is not used?

This seems like the edge case more than the standard.  If I'm reading JSON, I typically want the whole thing, and I want to know right away if it's going to fail.  Moreover, in an API setting (either web or library), I want to be able to inform the user/client right away if there's a problem and what it is.

**Comment by skyoxZ** [comment]
Close by #115856?

**Comment by eiriktsarpalis** [comment]
Good call, thanks. FYI @PranavSenthilnathan
